## A Meta-Analysis on the Transcriptomic Landscape of end-stage Human Heart Failure

### Abstract:

**Background**

During heart failure (HF) the myocardium undergoes ventricular remodeling, which is accompanied by changes in gene expression. Transcriptomic studies have contributed to fundamental knowledge of this remodeling. However, many studies do not agree on the differentially expressed marker genes. We hypothesized that a comprehensive meta-analysis of available studies would help to address this incongruence and provide a more unified biological understanding of the transcriptome in HF. 

**Methods**

We curated and uniformly processed 16 public transcriptomic data sets consisting of left ventricular samples from 263 healthy and 653 failing hearts. We assessed their comparability by applying classifiers based on transfer learning. We then combined all datasets to extract a consensus gene signature. Combining this consensus signature with biological prior knowledge, we estimated transcription factor and signalling pathway activities.

**Results**

Although single studies reported highly dissimilar marker genes, transfer knowledge based approaches revealed that disease patterns are conserved across studies. We derived a gene ranking that reflects consistent molecular hallmarks of HF. The ranking is strongly enriched in fibrosis related gene sets and contains numerous significantly altered transcription factors including Nanog, Sox2, Pbx3, Mef2; pathways, including Jak-Stat, and miRNAs, including MIR-206, MIR-18A.

**Discussion**

We demonstrated the feasibility of combining transcriptional studies from different technologies, years and centers. We extracted a consensus gene signature as a resource to understand common molecular patterns in HF. To our knowledge, this report represents the largest meta-analysis of transcriptional HF studies to date

### Graphical abstract

<img src="../www/graphical_abstract_wide.png" width="100%", align="middle">

### Citation
> TODO
